core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1289,24 +1308,50 @@ public void run() {
             channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
             haveBackends = true;
           }
-          final Map<String, ?> serviceConfig =
+
+          // Assuming no error in config resolution for now.
+          Map<String, ?> serviceConfig =
               config.get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG);
-          if (serviceConfig != null && !serviceConfig.equals(lastServiceConfig)) {
-            channelLogger.log(ChannelLogLevel.INFO, "Service config changed");
+          Attributes effectiveConfig = config;
+          if (!lookUpServiceConfig || serviceConfig == null) {
+            if (serviceConfig != null) { // lookUpServiceConfig is false
+              channelLogger.log(
+                  ChannelLogLevel.INFO,
+                  "Service config from name resolver discarded by channel settings");
+            }
+            if (defaultServiceConfig != null) {
+              channelLogger.log(ChannelLogLevel.INFO, "Using default service config");
+            } else if (lastServiceConfig != null && serviceConfig == null) {
+              // no default config, no current service config, last config is non-trivial
+              channelLogger.log(ChannelLogLevel.INFO, "Service config changed to null");
+            }
+
+            if (!(defaultServiceConfig == null && serviceConfig == null)) {
+              effectiveConfig = config.toBuilder()
```

Comment:
Even service config is already `null` in the attributes, it would make the implementation simpler to always update the attributes with the selected config (possibly null), because performance/allocation is not a concern for this code path. Please see my suggestion above.

Reply:
That will break an existing test (Attributes.EMPTY is not the same as attributes with a key and a null value), I might need to change the test.

Reply:
If it's just one test, please go ahead and update that test.

Reply:
There are 6 tests, each will need introduce an ArgumentCapture, so added a check `if (effectiveServiceConfig != serviceConfig)` instead.